recipes-support: Add recipes for Z3, CryptoMiniSat and Boolector STM solvers

ovpanait opened the pull request:
These SMT solvers can be used as backend for yosys-smtbmc to formally verify a design.

More information about all the SMT backends available for yosys-smtbmc can be found in the SymbiYosys readme page:
https://symbiyosys.readthedocs.io/en/latest/

ovpanait committed:
recipes-support: Add recipe for Z3 SMT solver

Z3 is a theorem prover from Microsoft Research. Z3 SMT solver can be
used in conjunction with yosys-smtbmc to formally verify a design. For
more details please see:

https://github.com/Z3Prover/z3/wiki
https://symbiyosys.readthedocs.io/en/latest/quickstart.html

Signed-off-by: Ovidiu Panait <ovpanait@gmail.com>

ovpanait committed:
recipes-support: Add recipe for CryptoMiniSat

CryptoMiniSat is an advanced incremental SAT solver compatible with
Boolector SMT solver. Boolector can be used in conjunction with
yosys-smtbmc to formally verify a design.

For more details on Boolector and yosys-smtbmc please see:
https://symbiyosys.readthedocs.io/en/latest/quickstart.html

Signed-off-by: Ovidiu Panait <ovpanait@gmail.com>

ovpanait committed:
recipes-support: Add recipe for Boolector SMT solver

Boolector is a Satisfiability Modulo Theories (SMT) solver for the
theories of fixed-size bit-vectors, arrays and uninterpreted functions.
Boolector can be used in conjunction with yosys-smtbmc to formally
verify a design.

For more details please see:
https://boolector.github.io/
https://symbiyosys.readthedocs.io/en/latest/quickstart.html

Signed-off-by: Ovidiu Panait <ovpanait@gmail.com>

nathanrossi commented on recipes-support/boolector/boolector_git.bb:
These should probably be a for loop.

ovpanait replied:
Will change to a for loop.

nathanrossi commented on recipes-support/boolector/btor2tools_git.bb:
Is the installed filename intentionally missing the `.h`?

ovpanait replied:
This will install btor2parser.h file to the ${D}/${includedir}/btor2parser directory, created earlier by install -d ${D}${includedir}/btor2parser.

The installed filename will still be btor2parser.h.

nathanrossi commented on recipes-support/cryptominisat/cryptominisat_git.bb:
It looks like this patch is not intended to be upstreamed?

Also it should be possible to generate the cache variable value and insert it into `toolchain.cmake` such that the patch is not required. Something like:

```
cmake_do_generate_toolchain_file_append() {
    cat >> ${WORKDIR}/toolchain.cmake <<EOF
set(HAVE__FPU_SETCW FALSE CACHE FORCE)
EOF
}
```

It could also be conditionally populated as true for i386/x86 via a recipe variable, if having it provides some sort of performance benefit.

ovpanait replied:
The patch is cross-compile specific. It removes the check_float_precision() call because it tries to run some binary on the build host to determine fpu capabilities. I didn't find a way yet to implement the check at build time, instead of running the binary on the host. So I don't have anything worth upstreaming.

I am therefore inclined to drop the patch and use the "cmake_do_generate_toolchain_file_append" approach instead.

nathanrossi reviewed (changes requested):
Please also add the three solvers to `fpga-toolkit` and `hdl-build-all`.

Additionally there appears to be a build failure for `cryptominisat-native`.

```
| CMake Error at cmsat5-src/cmake_install.cmake:64 (file):
|   file RPATH_CHANGE could not write new RPATH:
|
|     /home/nathan/build/hdl/tmp-glibc/work/x86_64-linux/cryptominisat-native/5.6.8+gitAUTOINC+e440d535cb-r0/recipe-sysroot-native/usr/lib
|
|   to the file:
|
|     /home/nathan/build/hdl/tmp-glibc/work/x86_64-linux/cryptominisat-native/5.6.8+gitAUTOINC+e440d535cb-r0/image/home/nathan/build/hdl/tmp-glibc/work/x86_64-linux/cryptominisat-native/5.6.8+gitAUTOINC+e440d535cb-r0/recipe-sysroot-native/usr/bin/cryptominisat5_simple
|
|   No valid ELF RPATH or RUNPATH entry exists in the file;
| Call Stack (most recent call first):
|   cmake_install.cmake:65 (include)
```

ovpanait commented:
Thanks for your review! I will work on the necessary changes and update this issue when ready.

ovpanait committed:
fpga-toolkit.bb: Add z3 and boolector SAT solvers to toolkit

Signed-off-by: Ovidiu Panait <ovpanait@gmail.com>

ovpanait committed:
hdl-build-all.bb: Add z3 and boolector to build targets

Signed-off-by: Ovidiu Panait <ovpanait@gmail.com>

ovpanait commented:
I have updated the patchset. I fixed the `cryptominisat-native` build failure and also added z3 and boolector to `fpga-toolkit` and `hdl-build-all`.

CryptoMiniSat is a backend used internally by boolector, so it can't be used with yosys-smtbmc directly. This is the reason why I left it out of `fpga-toolkit` and `hdl-build-all`.

nathanrossi commented:
Thanks!
